src/allocator.rs

```diff
@@ -279,6 +446,21 @@ impl Allocator {
                 let atom = self.atom_vec[idx as usize];
                 &self.u8_vec[atom.start as usize..atom.end as usize]
             }
+            (ObjectType::SmallAtom, val) => {
+                let len = len_for_value(val);
+                let mut idx = self.temp_idx.borrow_mut();
+                *self.temp_vec[*idx].borrow_mut() = val.to_be_bytes();
+                let ret = unsafe {
```

Comment:
Okay, so here's the ugly/scary part.

I wonder if you've thought about using 24 bits instead of 26 and then just using the storage for the index. Then you don't need this potentially scary temporary buffer hack.

Reply:
In fact, then you can repurpose the other two bits (26 vs 24) to be the length of the small atom, 0-3.

Reply:
I'm pretty sure you can make a lifetime `c'` that depends on *two* lifetimes `a'` and `b'`. So the lifetime of the resulting atom could depend on both the `Node` and the `Allocator`. If you can get this to work, it could solve 1.

Reply:
I suppose the 3rd problem would be byte order.

Reply:
Byte order would actually *not* be a problem, if you consider the 24 bits to be an atom byte string (of up to length 3) rather than a small integer.

This might change quite a few semantics about this patch, but I believe it might actually end up simpler since the two atom types would be much more similar since they would both be `&[u8]`. If you add a `visit_as_int`, you could better optimize the int conversion to a couple cpu instructions.

Reply:
it would change the trajectory of this change and veer away from making `BigInt`, `G1Element` and `G2Element` their own node types. Even though conversion to and from `u32` and `i32` is cheap, those types are expensive.

Reply:
I prefer the direction of moving away from the `atom()` API returning pointers